modules/nonce/replaycheck.go

```diff
@@ -0,0 +1,69 @@
+package nonce
+
+import (
+	"github.com/tendermint/basecoin"
+	"github.com/tendermint/basecoin/errors"
+	"github.com/tendermint/basecoin/stack"
+	"github.com/tendermint/basecoin/state"
+)
+
+//nolint
+const (
+	NameNonce = "nonce"
+)
+
+// ReplayCheck uses the sequence to check for replay attacks
+type ReplayCheck struct {
+	stack.PassOption
+}
+
+// Name of the module - fulfills Middleware interface
+func (ReplayCheck) Name() string {
+	return NameNonce
+}
+
+var _ stack.Middleware = ReplayCheck{}
+
+// CheckTx verifies tx is not being replayed - fulfills Middlware interface
+func (r ReplayCheck) CheckTx(ctx basecoin.Context, store state.KVStore,
+	tx basecoin.Tx, next basecoin.Checker) (res basecoin.Result, err error) {
+
+	stx, err := r.checkIncrementNonceTx(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	return next.CheckTx(ctx, store, stx)
+}
+
+// DeliverTx verifies tx is not being replayed - fulfills Middlware interface
+// NOTE It is okay to modify the sequence before running the wrapped TX because if the
+// wrapped Tx fails, the state changes are not applied
+func (r ReplayCheck) DeliverTx(ctx basecoin.Context, store state.KVStore,
+	tx basecoin.Tx, next basecoin.Deliver) (res basecoin.Result, err error) {
+
+	stx, err := r.checkIncrementNonceTx(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	return next.DeliverTx(ctx, store, stx)
+}
+
+// checkNonceTx varifies the nonce sequence, an increment sequence number
+func (r ReplayCheck) checkIncrementNonceTx(ctx basecoin.Context, store state.KVStore,
+	tx basecoin.Tx) (basecoin.Tx, error) {
+
+	// make sure it is a the nonce Tx (Tx from this package)
+	nonceTx, ok := tx.Unwrap().(Tx)
+	if !ok {
+		return tx, errors.ErrNoNonce()
+	}
+
+	// check the nonce sequence number
+	err := nonceTx.CheckIncrementSeq(ctx, store)
```

Comment:
I've gone back and forth on this over the years... whether the state changing actions should be a method on the tx that makes the change, or the handler that processes the tx.

I used to hate the redux method where the tx are dumb and the reducer knows how to handle the tx. But I have come across a number of cases where I want to use some state outside of the tx itself to control it.  Maybe you can initialize the Handler with `MaxSequence` and it will reject all nonce with a higher sequence....

It is easy to initialize this data on the handler, when the tx does the checking, how does it have access to this context?

I would inline `nonceTx.CheckIncrementSeq` right here.  But you don't have to.  We can talk more, these are subtle design patterns.